### ISSUE TITLE
Remove `caret` attribute from `<wa-button>`, `<wa-dropdown>` should add the caret automatically 

### ISSUE THREAD

**Issue by LeaVerou** [shoelace-style/webawesome#225]
This is a `<wa-button>` only thing:

```html
<wa-button size="small" caret>Small</wa-button>
<wa-button size="medium" caret>Medium</wa-button>
<wa-button size="large" caret>Large</wa-button>
```

produces
<img width="379" alt="image" src="https://github.com/user-attachments/assets/83e8ad50-f0ce-4628-a1c9-4157add5b789">

I would suggest we remove this one:
- It seems the main use case is to add a caret when using buttons together with `<wa-dropdown>`. But in that case, `<wa-dropdown>` should be adding the caret — the author intent has been made clear by using `<wa-dropdown>` and it creates an error condition in the current API that the author could forget to add it.
- Its purpose conflicts with the `suffix` slot. What happens if you specify both? Are there use cases for using them together?

**Comment by claviska** [comment]
> But in that case, <wa-dropdown> should be adding the caret — the author intent has been made clear by using <wa-dropdown> and it creates an error condition in the current API that the author could forget to add it.

Despite being a recommended pattern, users don't always want the caret to appear and there are some valid edge cases. Here's an example we show in the Breadcrumb docs where I agree that a caret doesn't make sense.

![CleanShot 2024-11-19 at 21 06 49@2x](https://github.com/user-attachments/assets/e2b63e1f-d2d6-4ba8-8cb7-b98f14129525)

Also, we try to avoid cross-contaminating component logic, although sometimes it's unavoidable (looking at you, button group!). In terms of components, it's generally considered a bad practice.

In the case of a button receiving a caret depending on its position in the DOM (as opposed to an attribute or a slotted icon, for example), I'm pretty sure that won't play nicely with SSR, but maybe @KonnorRogers can confirm/deny this assumption.

That said, I'm willing to look at a prototype if SSR isn't an issue, but I caution the cross-contamination of components and hope that we can keep things like that to a minimum and only when it significantly benefits UX.

> Its purpose conflicts with the suffix slot. What happens if you specify both? Are there use cases for using them together?

The `caret` attribute is basically sugar for slotting this system icon into the suffix:

```html
<wa-icon
  part="caret"
  class="button__caret"
  library="system"
  name="chevron-down"
  variant="solid"
></wa-icon>
```

When `caret` is used, the suffix slot is hidden by design. (This, however, appears to not be documented but [it _is_ intentional](https://github.com/shoelace-style/webawesome/blob/85c2a38cefe6fe9e898cfa1c6a7289454f024136/src/components/button/button.styles.ts#L333).)

**Comment by LeaVerou** [comment]
> > But in that case,  should be adding the caret — the author intent has been made clear by using  and it creates an error condition in the current API that the author could forget to add it.
> 
> Despite being a recommended pattern, users don't always want the caret to appear and there are some valid edge cases. Here's an example we show in the Breadcrumb docs where I agree that a caret doesn't make sense.
> 
> ![CleanShot 2024-11-19 at 21 06 49@2x](https://private-user-images.githubusercontent.com/55639/387900048-e2b63e1f-d2d6-4ba8-8cb7-b98f14129525.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIxMjUyOTAsIm5iZiI6MTczMjEyNDk5MCwicGF0aCI6Ii81NTYzOS8zODc5MDAwNDgtZTJiNjNlMWYtZDJkNi00YmE4LThjYjctYjk4ZjE0MTI5NTI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDE3NDk1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTlkNDg4ZjcxNzcwZDgwMmQzNTIzYmY3OTIxZTM5ZTc0YTVjNWE4OTNjNzUzNzQ2NzQ2MmY0ZjRkZmRiZjFjNDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.gvChB0fVEd7wrCXFBznqXHg-4hIiYq_TG5p0r9XLvUU)

Yup, there should be a way to hide it! But we shouldn't be optimizing for edge cases; the common case is the one that should be easiest/effortless.

> Also, we try to avoid cross-contaminating component logic, although sometimes it's unavoidable (looking at you, button group!). In terms of components, it's generally considered a bad practice.

Typically, you want good components to resemble native HTML as much as possible, except for the cases where diverging is a conscious decision to improve UX. For what it's worth, HTML itself does this a fair bit. Off the top of my head:
- A `<li>` renders completely differently inside a `<ul>` vs an `<ol>`
- An `<option>` renders differently based on whether it’s in a `<select>` or a `<datalist>`
- Until recently, headings were defined to have a different semantic level based on what sectioning elements they were inside. The only reason this changed was because browsers didn't implement it well, not because the pattern was deemed wrong.
- `<hr>` is a horizontal line in general, but a divider when used within a `<select>`.
- `<svg>` behaves differently based on whether it’s inside another `<svg>` or not
- `<th>` behaves differently in `<thead>` vs `<tbody>`

Presumably the reason this is considered a good practice is the [single responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle), and that's a good rule of thumb in general.
However:
- Remember the PoC: End-users > authors (our users) > implementors (that's us) > theoretical purity
- The code can be organized in a way so that all the dropdown-related stuff lives in dropdown. 
- I see it more as progressive enhancement than cross-contamination. Having rules that progressively enhance the UI created for certain element types is pretty common, we do it in WA too (e.g. badge renders differently in a button).
- It doesn’t even need to be `<wa-button>` specific. What if instead of using the suffix slot we add it in a way that works for any element, even `<button>`?

> In the case of a button receiving a caret depending on its position in the DOM (as opposed to an attribute or a slotted icon, for example), I'm pretty sure that won't play nicely with SSR, but maybe @KonnorRogers can confirm/deny this assumption.
> 
> That said, I'm willing to look at a prototype if SSR isn't an issue

This comes up frequently enough that I think we need to discuss it separately. I'll open an issue. [Edit: [done](https://github.com/shoelace-style/webawesome/discussions/228)]

**Comment by KonnorRogers** [comment]
> > But in that case,  should be adding the caret — the author intent has been made clear by using  and it creates an error condition in the current API that the author could forget to add it.
> 
> Despite being a recommended pattern, users don't always want the caret to appear and there are some valid edge cases. Here's an example we show in the Breadcrumb docs where I agree that a caret doesn't make sense.
> 
> ![CleanShot 2024-11-19 at 21 06 49@2x](https://private-user-images.githubusercontent.com/55639/387900048-e2b63e1f-d2d6-4ba8-8cb7-b98f14129525.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIxMzc5MjksIm5iZiI6MTczMjEzNzYyOSwicGF0aCI6Ii81NTYzOS8zODc5MDAwNDgtZTJiNjNlMWYtZDJkNi00YmE4LThjYjctYjk4ZjE0MTI5NTI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDIxMjAyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThhNzY3ZjEwOTczMDM4MTZkMzQ1ZmVhOWRlYzUxN2VhNjlhN2E5YzcyNmRkYTMwOWM2OWQwMWMxM2ZiYmQxZjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.-9B4Yt2q_QLm-3Jt3KTwCTsLo1atqvaYUPMSDM6iP3o)
> 
> Also, we try to avoid cross-contaminating component logic, although sometimes it's unavoidable (looking at you, button group!). In terms of components, it's generally considered a bad practice.
> 
> In the case of a button receiving a caret depending on its position in the DOM (as opposed to an attribute or a slotted icon, for example), I'm pretty sure that won't play nicely with SSR, but maybe @KonnorRogers can confirm/deny this assumption.
> 
> That said, I'm willing to look at a prototype if SSR isn't an issue, but I caution the cross-contamination of components and hope that we can keep things like that to a minimum and only when it significantly benefits UX.
> 
> > Its purpose conflicts with the suffix slot. What happens if you specify both? Are there use cases for using them together?
> 
> The `caret` attribute is basically sugar for slotting this system icon into the suffix:
> 
> ```
> <wa-icon
>   part="caret"
>   class="button__caret"
>   library="system"
>   name="chevron-down"
>   variant="solid"
> ></wa-icon>
> ```
> 
> When `caret` is used, the suffix slot is hidden by design. (This, however, appears to not be documented but [it _is_ intentional](https://github.com/shoelace-style/webawesome/blob/85c2a38cefe6fe9e898cfa1c6a7289454f024136/src/components/button/button.styles.ts#L333).)

in the strictest of senses, sure.

But in reality we should probably find a minimal DOM implementation for servers to be able to handle parent / child relationships.

Theres a few ideas here with using something like the "Context Proposal" or something like Linkedom.

Theres some questions I have around HTML streaming, but SSR feels like a "nice to have" as opposed to our holy grail at this point.

**Comment by claviska** [comment]
@KonnorRogers you know how I feel about SSR in general…a HUGE part of me is like "fuck SSR _(and, on a completely unrelated note, React)_" but I realize not having it is a show stopper for some of our potential customers.

I'm usually OK with just doing something that makes sense without worrying about SSR, but I also know how far you've come with your work on Web Awesome + SSR so I don't want to tread on that by doing more SSR-unfriendly things without considering it.

I know it's is still experimental in Web Awesome, but the SSR story only works if 99.9% or higher of it actually works. The more things we add that _don't_ work, the more janky it is and, well, _why did we even bother with it to begin with_ at that point?

Do you think this is something we have a reasonable chance at making work (without expending too much effort)? Or is this a bigger "industry" problem that needs to be fixed upstream?

**Comment by KonnorRogers** [comment]
> @KonnorRogers you know how I feel about SSR in general…a HUGE part of me is like "fuck SSR _(and, on a completely unrelated note, React)_" but I realize not having it is a show stopper for some of our potential customers.
> 
> I'm usually OK with just doing something that makes sense without worrying about SSR, but I also know how far you've come with your work on Web Awesome + SSR so I don't want to tread on that by doing more SSR-unfriendly things without considering it.
> 
> I know it's is still experimental in Web Awesome, but the SSR story only works if 99.9% or higher of it actually works. The more things we add that _don't_ work, the more janky it is and, well, _why did we even bother with it to begin with_ at that point?
> 
> Do you think this is something we have a reasonable chance at making work (without expending too much effort)? Or is this a bigger "industry" problem that needs to be fixed upstream?

my current thoughts are figuring out something with Lit's SSR module and how to handle the common cases like:

parent / child relationships and query selectors. Those seem to be the biggest problems with the current SSR solution since components dont have access to the full DOM tree, they only know about themselves. I imagine there may be other small bits and pieces to worry about, but those seem to be the main offenders.

I _feel_ like things like this which only need access to the tree and aren't doing funky thing like window / container measurements or anything that would truly need a browser feel like fair game to me.

**Comment by LeaVerou** [comment]
> @KonnorRogers you know how I feel about SSR in general…a HUGE part of me is like "fuck SSR _(and, on a completely unrelated note, React)_" but I realize not having it is a show stopper for some of our potential customers.
> 
> I'm usually OK with just doing something that makes sense without worrying about SSR, but I also know how far you've come with your work on Web Awesome + SSR so I don't want to tread on that by doing more SSR-unfriendly things without considering it.
> 
> I know it's is still experimental in Web Awesome, but the SSR story only works if 99.9% or higher of it actually works. The more things we add that _don't_ work, the more janky it is and, well, _why did we even bother with it to begin with_ at that point?
> 
> Do you think this is something we have a reasonable chance at making work (without expending too much effort)? Or is this a bigger "industry" problem that needs to be fixed upstream?

I opened https://github.com/shoelace-style/webawesome/discussions/228 so we can discuss this more generally so that we have a general framework and we don't have to decide on a case by case basis (we'd still need to consider if each case warrants an exception, but that's an easier decision).

But it’s not about whether it works, IMO if we have syntax to support SSR, even if some shortcuts don't work in SSR, then we support SSR, period. But everything we do needs to be _possible_ in an SSR environment, even if it's less _easy_.

**Comment by claviska** [comment]
> even if some shortcuts don't work in SSR

Users that demand SSR do not want jank. Period. They're not used to it. They'll complain.

All I'm trying to say here is that, unless we can guarantee something like `with-caret` will work properly (i.e. with no jank), it immediately becomes a liability for every SSR customer.

**Comment by LeaVerou** [comment]
Maybe I'm misunderstanding what you mean by jank, but I completely agree the end result needs to be equally good. What I was saying is, I think it's okay if they need to write more code to accomplish the same result, assuming they understand why. Do you think that's not accurate?

**Comment by claviska** [comment]
I agree with that. I’m unsure if we can accurately measure the caret in an SSR environment, which would result in jank (shifting/movement on hydrate).

We probably can _right now_ since size maps to a specific font size and height for buttons. If we move to scalable buttons/inputs, the story might be different.

**Comment by LeaVerou** [comment]
> All I'm trying to say here is that, unless we can guarantee something like `with-caret` will work properly

Wrt `with-caret` (in case it was not just an example): IMO we should make the common case easy, and the common case is to want an arrow, right? So I think the component should add an arrow anyway, and provide a way to opt out. Dropdown doesn't even need to special case buttons: it can check if the trigger element has a `suffix` slot and use that if so, otherwise use `::after` or something. 

Now, *customizing* the icon seems like a common need. Perhaps we could have a dual purpose attribute that either hides the icon, or changes it. Though I suspect you'd more frequently want to customize it across all/many dropdowns, so having to specify something like `suffix="caret-down"` over and over is not great DX either… 🤔 

> I agree with that. I’m unsure if we can accurately measure the caret in an SSR environment, which would result in jank (shifting/movement on hydrate).
> 
> We probably can _right now_ since size maps to a specific font size and height for buttons. If we move to scalable buttons/inputs, the story might be different.

Agreed that that kind of jank would not be acceptable. What I'm missing is, why do we need to know its size to prevent jank?

**Comment by KonnorRogers** [comment]
> > even if some shortcuts don't work in SSR
> 
> Users that demand SSR do not want jank. Period. They're not used to it. They'll complain.
> 
> All I'm trying to say here is that, unless we can guarantee something like `with-caret` will work properly (i.e. with no jank), it immediately becomes a liability for every SSR customer.

fwiw, svgs that are fetched dont work on the server today, we just kinda "guess" a 16x16 svg is "good" enough to prevent layout shifting.

To me, the thing that makes the most sense in my head is to leave the `caret` attribute on buttons.

When a `<wa-dropdown>` has a "slotchange" and it detects a `<wa-button>` as a trigger, it adds the `caret` attribute to `<wa-button>`

The obvious caveat is this will not play nicely with SSR / frontend frameworks and anything doing "morphing" / "dom diffing", so we'd need to document for them to add the `caret` attribute.

I guess another idea would be having the caret live inside of the `<wa-dropdown>` shadow dom and always render it next to whatever button / element is slotted.

It can accept a "slot" for a custom caret, or be disabled with something like `hide-caret`

**Comment by claviska** [comment]
> I guess another idea would be having the caret live inside of the <wa-dropdown> shadow dom and always render it next to whatever button / element is slotted.

this won’t work because the caret is inside the button’s shadow root — unless we remove caret and slot it into the button as a suffix…

**Comment by LeaVerou** [comment]
Btw #234 can improve on this before we need to figure out what's the best way we can auto-add it. At least then we can just recommend `icon-end="caret"` (or whatever the icon id is) rather than a whole separate attribute that is [overfitted](https://medium.com/design-bootcamp/overfitting-and-the-problem-with-use-cases-337d9f4bf4d7) to this particular use case . And #447 can solve this completely: dropdown can just set `--wa-icon-end: caret` to `slot[name=trigger]::slotted(*)`, which will inherit with no undesirable couplings.
